src/main/java/com/uber/cadence/internal/worker/autoscaler/PollerAutoScaler.java

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.uber.cadence.internal.worker.autoscaler;
+
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PollerAutoScaler {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PollerAutoScaler.class);
+
+  private Duration coolDownTime;
+  private PollerUsageEstimator pollerUsageEstimator;
+  private Recommender recommender;
+  private ResizableSemaphore semaphore;
+  private int semaphoreSize;
+  private boolean shutingDown;
+
+  public PollerAutoScaler(
+      Duration coolDownTime, PollerUsageEstimator pollerUsageEstimator, Recommender recommender) {
+    this.coolDownTime = coolDownTime;
+    this.pollerUsageEstimator = pollerUsageEstimator;
+    this.recommender = recommender;
+    this.semaphore = new ResizableSemaphore(recommender.getUpperValue());
+    this.semaphoreSize = recommender.getUpperValue();
+  }
+
+  public void start() {
+    Executors.newSingleThreadExecutor()
+        .submit(
+            new Runnable() {
+              @Override
+              public void run() {
+                while (!shutingDown) {
+                  try {
+                    Thread.sleep(coolDownTime.toMillis());
+                    if (!shutingDown) {
+                      resizePollers();
+                    }
+                  } catch (InterruptedException e) {
```

Comment:
do you need to handle anything here (maybe logging)

Reply:
Don't think it is necessary, but might add some log just for obervability.